File: examples/python/03_batch_api.py
"""
03 - Batch API end-to-end (with caching)
========================================

Demonstrates the cheapest async pattern available on Claude:
- Submit up to 10,000 requests in one batch
- 50% discount on input AND output tokens
- Use 1-hour cache TTL because batches can take >5 minutes

This script:
1. Builds a batch of 5 requests, each with a shared cached system prompt.
2. Submits them.
3. Polls until done.
4. Prints results with usage breakdown.

Run:
    python 03_batch_api.py

Note: most batches finish in under a minute, but the SLA is 24 hours.
Don't use the Batch API for anything user-facing in real time.
"""

from __future__ import annotations

import os
import sys
import time

import anthropic
from anthropic.types.message_create_params import MessageCreateParamsNonStreaming
from anthropic.types.messages.batch_create_params import Request

MODEL = "claude-haiku-4-5"   # cheap; the discount stacks regardless of model

# Long enough to clear the 1,024-token cache minimum.
SYSTEM_PROMPT = """\
You are a content classifier. Read the input text and output a JSON object
with these fields and nothing else:
{
  "category": one of ["billing", "bug", "feature_request", "praise", "other"],
  "sentiment": one of ["positive", "neutral", "negative"],
  "urgency": integer 1-5,
  "summary": one sentence under 25 words
}

Rules:
- Always output valid JSON only - no commentary, no markdown fences.
- If the message is empty, return all fields as null.
- Choose the SINGLE best category, even if it's a stretch.
- Urgency 5 means user is blocked or business is at risk.
- Be terse in the summary; do not editorialise.
""" * 5


SAMPLE_INPUTS = [
    "I was double-charged for last month and need this refunded ASAP.",
    "Your new export feature is brilliant - saved me hours this week!",
    "Can you add CSV export to the reports page? It's the only thing missing.",
    "The dashboard hangs whenever I filter by 'last 90 days'. Tried 3 browsers.",
    "Hi, just wondering if there's a student discount.",
]


def build_requests() -> list[Request]:
    requests: list[Request] = []
    for i, text in enumerate(SAMPLE_INPUTS):
        requests.append(
            Request(
                custom_id=f"classify-{i:04d}",
                params=MessageCreateParamsNonStreaming(
                    model=MODEL,
                    max_tokens=200,
                    system=[
                        {
                            "type": "text",
                            "text": SYSTEM_PROMPT,
                            "cache_control": {"type": "ephemeral", "ttl": "1h"},
                        }
                    ],
                    messages=[{"role": "user", "content": text}],
                ),
            )
        )
    return requests


def main() -> int:
    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        print("Set ANTHROPIC_API_KEY", file=sys.stderr)
        return 1

    client = anthropic.Anthropic(api_key=api_key)

    print("Submitting batch...")
    batch = client.messages.batches.create(requests=build_requests())
    print(f"  batch id = {batch.id}")
    print(f"  status   = {batch.processing_status}")

    print("\nPolling (most batches finish in <1 minute)...")
    while batch.processing_status != "ended":
        time.sleep(15)
        batch = client.messages.batches.retrieve(batch.id)
        counts = batch.request_counts
        print(
            f"  status={batch.processing_status} "
            f"processing={counts.processing} succeeded={counts.succeeded} "
            f"errored={counts.errored}"
        )

    print("\nResults:")
    total_input = total_output = total_cache_create = total_cache_read = 0
    for result in client.messages.batches.results(batch.id):
        if result.result.type != "succeeded":
            print(f"  [{result.custom_id}] FAILED: {result.result.type}")
            continue
        msg = result.result.message
        u = msg.usage
        total_input        += u.input_tokens
        total_output       += u.output_tokens
        total_cache_create += u.cache_creation_input_tokens or 0
        total_cache_read   += u.cache_read_input_tokens or 0
        print(f"  [{result.custom_id}] -> {msg.content[0].text}")

    print("\nAggregate usage across the batch:")
    print(f"  fresh_input          = {total_input}")
    print(f"  cache_creation_input = {total_cache_create}")
    print(f"  cache_read_input     = {total_cache_read}")
    print(f"  output               = {total_output}")
    print(
        "\nNote: Batch API rates are 50% of standard. "
        "When you compute cost, multiply each line above by the corresponding "
        "Batch rate, not the sync rate. See 07_cost_calculator.py."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/typescript/src/03_batch_api.ts
/**
 * 03 - Batch API end-to-end (with caching)
 * ========================================
 *
 * - Submit up to 10,000 requests in one batch
 * - 50% discount on input AND output tokens
 * - Use 1-hour cache TTL because batches can take >5 minutes
 *
 * Run:
 *     npm run example:03
 *
 * Most batches finish in <1 minute; SLA is 24 hours.
 */

import "dotenv/config";
import Anthropic from "@anthropic-ai/sdk";

const MODEL = "claude-haiku-4-5";

const SYSTEM_PROMPT =
  `You are a content classifier. Read the input text and output a JSON object
with these fields and nothing else:
{
  "category": one of ["billing", "bug", "feature_request", "praise", "other"],
  "sentiment": one of ["positive", "neutral", "negative"],
  "urgency": integer 1-5,
  "summary": one sentence under 25 words
}

Rules:
- Always output valid JSON only - no commentary, no markdown fences.
- If the message is empty, return all fields as null.
- Choose the SINGLE best category, even if it's a stretch.
- Urgency 5 means user is blocked or business is at risk.
- Be terse in the summary; do not editorialise.
`.repeat(5);

const SAMPLE_INPUTS = [
  "I was double-charged for last month and need this refunded ASAP.",
  "Your new export feature is brilliant - saved me hours this week!",
  "Can you add CSV export to the reports page? It's the only thing missing.",
  "The dashboard hangs whenever I filter by 'last 90 days'. Tried 3 browsers.",
  "Hi, just wondering if there's a student discount.",
];

async function main(): Promise<void> {
  if (!process.env.ANTHROPIC_API_KEY) {
    console.error("Set ANTHROPIC_API_KEY");
    process.exit(1);
  }
  const client = new Anthropic();

  const requests = SAMPLE_INPUTS.map((text, i) => ({
    custom_id: `classify-${String(i).padStart(4, "0")}`,
    params: {
      model: MODEL,
      max_tokens: 200,
      system: [
        {
          type: "text" as const,
          text: SYSTEM_PROMPT,
          cache_control: { type: "ephemeral" as const, ttl: "1h" as const },
        },
      ],
      messages: [{ role: "user" as const, content: text }],
    },
  }));

  console.log("Submitting batch...");
  let batch = await client.messages.batches.create({ requests });
  console.log(`  batch id = ${batch.id}`);
  console.log(`  status   = ${batch.processing_status}`);

  console.log("\nPolling (most batches finish in <1 minute)...");
  while (batch.processing_status !== "ended") {
    await new Promise((r) => setTimeout(r, 15_000));
    batch = await client.messages.batches.retrieve(batch.id);
    const c = batch.request_counts;
    console.log(
      `  status=${batch.processing_status} processing=${c.processing} succeeded=${c.succeeded} errored=${c.errored}`,
    );
  }

  console.log("\nResults:");
  let totalInput = 0,
    totalOutput = 0,
    totalCacheCreate = 0,
    totalCacheRead = 0;

  for await (const result of client.messages.batches.results(batch.id)) {
    if (result.result.type !== "succeeded") {
      console.log(`  [${result.custom_id}] FAILED: ${result.result.type}`);
      continue;
    }
    const msg = result.result.message;
    const u = msg.usage;
    totalInput += u.input_tokens;
    totalOutput += u.output_tokens;
    totalCacheCreate += u.cache_creation_input_tokens ?? 0;
    totalCacheRead += u.cache_read_input_tokens ?? 0;
    const t = msg.content.find((b) => b.type === "text");
    const text = t?.type === "text" ? t.text : "(no text)";
    console.log(`  [${result.custom_id}] -> ${text}`);
  }

  console.log("\nAggregate usage across the batch:");
  console.log(`  fresh_input          = ${totalInput}`);
  console.log(`  cache_creation_input = ${totalCacheCreate}`);
  console.log(`  cache_read_input     = ${totalCacheRead}`);
  console.log(`  output               = ${totalOutput}`);
  console.log(
    "\nNote: Batch API rates are 50% of standard. " +
      "Multiply each line above by the corresponding Batch rate. " +
      "See 07_cost_calculator.",
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
